data_structures/trie_using_hashmap.cpp

```diff
@@ -0,0 +1,345 @@
+    bool startwith(const std::string& prefix) {
+        std::shared_ptr<Node> curr = root_node;
+        for (char ch : prefix) {
+            if (curr->children.find(ch) == curr->children.end()) {
+                return false;
+            }
+            curr = curr->children[ch];
+        }
+        return true;
+    }
+
+    /**
+     * @brief delete a word/string from a trie
+     * @param word string to delete from trie
+     */
+    void delete_word(std::string word) {
+        std::shared_ptr<Node> curr = root_node;
+        std::stack<std::shared_ptr<Node>> nodes;
+        int cnt = 0;
+        for (char ch : word) {
+            if (curr->children.find(ch) == curr->children.end()) {
+                return;
+            }
+            if (curr->word_end) {
+                cnt++;
+            }
+
+            nodes.push(curr->children[ch]);
+            curr = curr->children[ch];
+        }
+        // Delete only when it's a word, and it has children after
+        // or prefix in the line
+        if (nodes.top()->word_end) {
+            nodes.top()->word_end = false;
+        }
+        // Delete only when it has no children after
+        // and also no prefix in the line
+        while (!(nodes.top()->word_end) && nodes.top()->children.empty()) {
+            nodes.pop();
+            nodes.top()->children.erase(word.back());
+            word.pop_back();
+        }
+    }
+
+    /**
+     * @brief helper function to predict/recommend words that starts with a
+     * given prefix from the end of prefix's node iterate through all the child
+     * nodes by recursively appending all the possible words below the trie
+     * @param prefix string to recommend the words
+     * @param element node at the end of the given prefix
+     * @param results list to store the all possible words
+     * @returns list of recommended words
+     */
+    std::vector<std::string> get_all_words(std::vector<std::string> results,
+                                           const std::shared_ptr<Node>& element,
+                                           std::string prefix) {
+        if (element->word_end) {
+            results.push_back(prefix);
+        }
+        if (element->children.empty()) {
+            return results;
+        }
+        for (auto const& x : element->children) {
+            std::string key = "";
+            key = x.first;
+            prefix += key;
+
+            results =
+                get_all_words(results, element->children[x.first], prefix);
+
+            prefix.pop_back();
+        }
+
+        return results;
+    }
+
+    /**
+     * @brief predict/recommend a word that starts with a given prefix
+     * @param prefix string to search for
+     * @returns list of recommended words
+     */
+    std::vector<std::string> predict_words(const std::string& prefix) {
+        std::vector<std::string> result;
+        std::shared_ptr<Node> curr = root_node;
+        // treversing till the end of given prefix in trie
```

Comment:
Shouldn't here be `traversing`?
```suggestion
        // traversing until the end of the given prefix in trie
```

Reply:
yes it's traversing, I will update it. Thank you